chain/chain/src/resharding/manager.rs

```diff
@@ -84,34 +82,108 @@ impl ReshardingManager {
+        match resharding_event_type {
+            Some(ReshardingEventType::SplitShard(split_shard_event)) => {
+                self.split_shard(
+                    chain_store_update,
+                    block,
+                    shard_uid,
+                    tries,
+                    split_shard_event,
+                    next_shard_layout,
+                )?;
+            }
+            None => {
+                tracing::debug!(target: "resharding", ?resharding_event_type, "unsupported resharding event type, skipping");
+            }
         };
+        Ok(())
+    }
+
+    fn split_shard(
+        &mut self,
+        chain_store_update: ChainStoreUpdate,
+        block: &Block,
+        shard_uid: ShardUId,
+        tries: ShardTries,
+        split_shard_event: ReshardingSplitShardParams,
+        next_shard_layout: ShardLayout,
+    ) -> Result<(), Error> {
         if split_shard_event.parent_shard != shard_uid {
             let parent_shard = split_shard_event.parent_shard;
             tracing::debug!(target: "resharding", ?parent_shard, "shard uid does not match event parent shard, skipping");
             return Ok(());
         }
 
-        // TODO(resharding): what if node doesn't have memtrie? just pause
-        // processing?
-        // TODO(resharding): fork handling. if epoch is finalized on different
-        // blocks, the second finalization will crash.
-        tries.freeze_mem_tries(
+        // Reshard state by setting shard UId mapping from children to parent.
+        self.set_state_shard_uid_mapping(&split_shard_event)?;
+
+        // Create temporary children memtries by freezing parent memtrie and referencing it.
+        self.process_memtrie_resharding_storage_update(
+            chain_store_update,
+            block,
             shard_uid,
-            vec![split_shard_event.left_child_shard, split_shard_event.right_child_shard],
+            tries,
+            split_shard_event.clone(),
         )?;
 
         // Trigger resharding of flat storage.
         self.flat_storage_resharder.start_resharding(
-            ReshardingEventType::SplitShard(split_shard_event.clone()),
+            ReshardingEventType::SplitShard(split_shard_event),
             &next_shard_layout,
         )?;
 
-        let chunk_extra = self.get_chunk_extra(block_hash, &shard_uid)?;
+        Ok(())
+    }
+
+    /// Store in the database the mapping of shard UId from children to the parent shard,
+    /// so that subsequent accesses to the State will use the parent shard's UId as a prefix for the database key.
+    fn set_state_shard_uid_mapping(
+        &mut self,
+        split_shard_event: &ReshardingSplitShardParams,
+    ) -> io::Result<()> {
+        let mut store_update = self.store.trie_store().store_update();
+        let parent_shard_uid = split_shard_event.parent_shard;
+        // TODO(reshardingV3) No need to set the mapping for children shards that we won't track just after resharding?
```

Comment:
Good call making that a todo. Let's come back to this question when we figure out the full picture for post resharding cleanup.